keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+
+Once compiled for Windows, container images based on Window Base images (like NanoServer) needs to be published and maintained.
+
+New YAMLs for DaemonSets associated with CSI Node Plugins needs to be authored that will (1) target Windows nodes and (2) use Windows paths for domain socket related paths as illustrated in the User Story section above.
+
+### Risks and Mitigations
+
+Any pod on a Windows node can be configured to mount `\\.\pipe\csi-proxy` and perform privileged operations. Thus csi-proxy presents a potential security risk. To mitigate the risk, an admission webhook will need to be implemented and deployed that will reject all containers that bind mount `\\.\pipe\csi-proxy` as a hostPath volume but does not have `privileged` flag set in the pod's securityContext specification. This allows the `privileged` setting to be used for Windows pods as an indication the container will perform privileged operations. Other cluster-wide policies (e.g. PSP) that act on the `privileged` setting in a container's `securityContext` can enforce the same for CSI Node plugins targeting Windows nodes.
+
+## Design Details
+
+### Test Plan
+
+Unit tests will be added to verify Windows related enhancements in existing Kubernetes components mentioned above.
+
+All E2E storage tests covering CSI plugins will be ported to Windows workloads and successfully executed with above enhancements in place along with csi-proxy.exe.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- CSI Node Plugins for GCE PD, Azure Disk and Azure File can support Windows using csi-proxy.exe.
+- end-2-end tests in place involving CSI storage scenarios that exercise csi-proxy.exe and related enhancements.
+
+#### Beta -> GA Graduation
+
+- All in-tree storage plugins that implements support for Windows can use csi-proxy.exe along with other enhancements listed above to successfully deploy CSI plugins on Windows nodes.
+- Successful usage of csi-proxy.exe in Windows nodes by at least two storage vendors.
+
+### Upgrade / Downgrade Strategy
+
+If a cluster is downgraded from a version of Kubernetes where the plugin watcher supports Windows nodes to one that does not, existing Windows workloads that were using CSI plugins to access storage will no longer have access to the data. This loss of functionality cannot be handled in an elegant fashion.
+
+### Version Skew Strategy
+
+Different nodes in the cluster may be configured with different versions of csi-proxy.exe as part of a rolling upgrade of csi-proxy.exe. In such a scenario, it is recommended that csi-proxy.exe is upgraded first and once that is complete, the CSI Node Plugins that can take advantage of the new version of csi-proxy.exe is rolled out.
```

Comment:
How would the failure caused by the version skew be manifested to the users?

Reply:
Will call out explicitly that plugins need to check version of the proxy using the version API call and fail initialization the plugin detects an incompatibility.